python/lsst/daf/butler/_butler.py

```diff
@@ -660,7 +660,10 @@ def remove(self, datasetRefOrType, dataId=None, *, delete=True, remember=True, *
                 self.datastore.remove(ref)
             elif ref.isComposite():
                 for r in ref.components.values():
-                    self.datastore.remove(r)
+                    # If a dataset was removed previously but remembered
+                    # in registry, skip the removal in the datastore.
+                    if self.registry.getDatasetLocations(r):
+                        self.datastore.remove(r)
```

Comment:
Note that this attempts to remove the dataset from this particular datastore if it is present in any datastore.  I think that's okay for now, but we should open a ticket to address it in the future.  The more I think about it, the more I think we'd be better off if butler had a dict of datastores instead of having `ChainedDatastore`, so it could refer to them by name in this and other places.

Reply:
This is fine so long as the datastores used by the registry haven't changed because of a config change. I can filter the returned list by asking for the equivalent from the datastore itself.

Reply:
I think I've fixed it now so that it does use the datastore name before deciding whether to remove.